File: Cpp/Dijkstra.c
#include <stdbool.h>
#include <stdio.h>
#include <limits.h>
#include "Graph.h"

void dijkstra(Graph* g, int src) {
    int dist[MAX_VERTICES];
    bool sptSet[MAX_VERTICES];

    for (int i = 0; i < g->V; i++) {
        dist[i] = INT_MAX;
        sptSet[i] = false;
    }

    dist[src] = 0;

    for (int count = 0; count < g->V - 1; count++) {
        int u = -1;
        int minDist = INT_MAX;

        for (int v = 0; v < g->V; v++) {
            if (!sptSet[v] && dist[v] < minDist) {
                u = v;
                minDist = dist[v];
            }
        }

        if (u == -1) {
            break; // No reachable vertices left
        }

        sptSet[u] = true;

        for (int v = 0; v < g->V; v++) {
            if (!sptSet[v] && g->adj[u][v] && dist[u] != INT_MAX &&
                dist[u] + g->weight[u][v] < dist[v]) {
                dist[v] = dist[u] + g->weight[u][v];
            }
        }
    }

    printf("Shortest distances from source vertex %d:\n", src);
    for (int i = 0; i < g->V; i++) {
        printf("To vertex %d: %d\n", i, dist[i]);
    }
}

File: Python/Dijkstra.py
import heapq

def dijkstra(graph, start):
    # Priority queue to store vertices and their tentative distances
    priority_queue = [(0, start)]  # (distance, vertex)
    distances = {vertex: float('infinity') for vertex in graph}
    distances[start] = 0

    while priority_queue:
        current_distance, current_vertex = heapq.heappop(priority_queue)

        # Skip if the current distance is greater than the known distance
        if current_distance > distances[current_vertex]:
            continue

        # Update distances for neighbors
        for neighbor, weight in graph[current_vertex].items():
            distance = current_distance + weight

            # If a shorter path is found, update the distance
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))

    return distances

# Example usage:
graph = {
    'A': {'B': 1, 'C': 4},
    'B': {'A': 1, 'C': 2, 'D': 5},
    'C': {'A': 4, 'B': 2, 'D': 1},
    'D': {'B': 5, 'C': 1}
}

start_vertex = 'A'
shortest_distances = dijkstra(graph, start_vertex)

print(f"Shortest distances from vertex {start_vertex}:")
for vertex, distance in shortest_distances.items():
    print(f"To {vertex}: {distance}")
